### PR TITLE
feat(swapper): implement getBestSwapper logic

### PR DESCRIPTION
**Pull request by 0xApotheosis**
Contributes to https://github.com/shapeshift/web/issues/3466

### CODE REVIEW

File: packages/swapper/src/manager/SwapperManager.ts [hunk 1]
```diff
@@ -62,10 +73,40 @@ export class SwapperManager {
...
+    })
+
+    // Get quotes from all swappers that support the pair
+    const quotePromises: Promise<SwapperQuoteTuple>[] = supportedSwappers.map(
+      async (swapper) => [swapper, await swapper.getTradeQuote(args)] as const,
+    )
+    const settledQuoteRequests = await Promise.allSettled(quotePromises)
+    // For each swapper, get output amount/(input amount + gas fee), where all values are in fiat
+    const fulfilledQuoteTuples = settledQuoteRequests
+      .filter(isFulfilled)
```
Comment (gomesalexandre) [inline packages/swapper/src/manager/SwapperManager.ts:92]: 🧠

File: packages/swapper/src/manager/SwapperManager.test.ts [hunk 2]
```diff
@@ -1,50 +1,21 @@
...
+  badTradeQuote,
+  cowSwapper,
+  goodTradeQuote,
+  thorchainSwapper,
+  zrxAvalancheSwapper,
+  zrxEthereumSwapper,
+} from './testData'
 
 describe('SwapperManager', () => {
-  const zrxEthereumSwapperDeps: ZrxSwapperDeps = {
```
Comment (0xApotheosis) [inline packages/swapper/src/manager/SwapperManager.test.ts:19]: Moved into `testData.ts` for reusability.

File: packages/swapper/src/manager/SwapperManager.ts [hunk 3]
```diff
@@ -62,10 +73,50 @@ export class SwapperManager {
     return this
   }
 
-  async getBestSwapper(args: ByPairInput): Promise<Swapper<ChainId> | undefined> {
-    // TODO: This will eventually have logic to determine the best swapper.
-    // For now we return the first swapper we get from getSwappersByPair
-    return this.getSwappersByPair(args)[0]
+  async getBestSwapper(args: GetBestSwapperArgs): Promise<Swapper<ChainId> | undefined> {
```
Comment (0xApotheosis) [inline packages/swapper/src/manager/SwapperManager.ts:82]: This is where the magic happens.

File: packages/swapper/src/manager/utils.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { Asset } from '@shapeshiftoss/asset-service'
+import { ChainId } from '@shapeshiftoss/caip'
+
+import { Swapper, TradeQuote } from '../api'
+import { bnOrZero, fromBaseUnit } from '../swappers/utils/bignumber'
+
+export const getRatioFromQuote = async (
```
Comment (0xApotheosis) [inline packages/swapper/src/manager/utils.ts:19]: This function is the crucial piece of logic in this PR.

**Review by 0xApotheosis — COMMENTED** [review]
Comments for reviewers.

File: packages/swapper/src/manager/SwapperManager.ts (hunk 3 above)
Comment (0xdef1cafe) [reply packages/swapper/src/manager/SwapperManager.ts:82]: popping this here for reference https://twitter.com/0xngmi/status/1610492831134629889?s=20&t=hBUmFYpEFEwNxQvLj1Fbag

File: packages/swapper/src/manager/testData.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+import { ZrxSwapperDeps } from '../swappers/zrx/types'
+
+const zrxEthereumSwapperDeps: ZrxSwapperDeps = {
+  web3: <Web3>{},
+  adapter: <ethereum.ChainAdapter>{
+    getChainId: () => KnownChainIds.EthereumMainnet,
+  },
+}
+
+export const zrxEthereumSwapper = new ZrxSwapper(zrxEthereumSwapperDeps)
```
Comment (0xdef1cafe) [inline packages/swapper/src/manager/testData.ts:23]: is this going to get included into the build, and hence instantiate swappers at runtime in web?

File: packages/swapper/src/manager/utils.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  feeAsset: Asset,
+): Promise<number> => {
+  /*
+    We make an assumption here that swappers return comparable asset rates.
+    If a swapper were to undervalue the outbound asset and overvalue the inbound asset
+    then the ratio would be incorrectly in its favour.
+  */
+  const sellAssetUsdRate = await swapper.getUsdRate(quote.sellAsset)
+  const buyAssetUsdRate = await swapper.getUsdRate(quote.buyAsset)
+  const feeAssetUsdRate = await swapper.getUsdRate(feeAsset)
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply packages/swapper/src/manager/utils.ts]: Addressed in https://github.com/shapeshift/lib/pull/1123/commits/efe3b772451b1212b8b90b6ca7af396694e43f2c.

File: packages/swapper/src/manager/testData.ts (hunk 5 above)
Comment (0xApotheosis) [reply packages/swapper/src/manager/testData.ts]: Indeed. Fixed up in https://github.com/shapeshift/lib/pull/1123/commits/7d9fc7531fae1c585c60d37845774f144f0ecbdc.

File: packages/swapper/src/manager/utils.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  const totalReceiveAmountFiat = bnOrZero(
+    fromBaseUnit(quote.buyAmountCryptoBaseUnit, quote.buyAsset.precision),
+  )
+    .times(buyAssetUsdRate)
+    .minus(bnOrZero(quote.feeData.buyAssetTradeFeeUsd))
+  const networkFeeFiat = bnOrZero(
+    fromBaseUnit(quote.feeData.networkFeeCryptoBaseUnit, feeAsset.precision),
+  ).times(feeAssetUsdRate)
+
+  return totalReceiveAmountFiat.div(totalSellAmountFiat.plus(networkFeeFiat)).toNumber()
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply packages/swapper/src/manager/utils.ts]: Handled in https://github.com/shapeshift/lib/pull/1123/commits/88e264735809aead1ddfb20aeed859514b1bfb62.

It's a bit defensive, but safety first.

**Review by 0xdef1cafe — APPROVED** [review]
very naice

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**